### PR TITLE
MM-47079: Migrate "components/suggestion/search_user_provider.jsx" to Typescript

### PR DESCRIPTION
**Pull request by babinderrathi**
#### Summary
Migrate of search_user_provider.jsx" to Typescript

#### Ticket Link
 Fixes https://github.com/mattermost/mattermost-server/issues/21090
 Fixes https://mattermost.atlassian.net/browse/MM-47079

#### Related Pull Requests
None

#### Screenshots
none

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: components/suggestion/search_user_provider.tsx [hunk 1]
```diff
@@ -75,20 +81,21 @@ class SearchUserSuggestion extends Suggestion {
 }
 
 export default class SearchUserProvider extends Provider {
-    constructor(userSearchFunc) {
+    autocompleteUsersInTeam: (username: string) => any;
```
Comment (babinderrathi) [inline components/suggestion/search_user_provider.tsx:84]: Due to type mismatching, I am unable to find any solution which can replace 'any' here .

File: components/suggestion/search_user_provider.tsx [hunk 2]
```diff
@@ -75,20 +81,21 @@ class SearchUserSuggestion extends Suggestion {
...
+    constructor(userSearchFunc: (username: string) => DispatchFunc) {
```
Comment (sinansonmez) [inline components/suggestion/search_user_provider.tsx:85]: If you track back `userSearchFunc` func, you will see that it is returning `UserAutocomplete` as seen below.

https://github.com/mattermost/mattermost-webapp/blob/7d527c9c852f9e931cf253c16f663371340738d5/packages/client/src/client4.ts#L913-L916

Therefore, this change should do the trick. 
```suggestion
    autocompleteUsersInTeam: (username: string) => Promise<UserAutocomplete>;
    constructor(userSearchFunc: (username: string) => Promise<UserAutocomplete>) {
```

You also need to change the return type of `autocompleteUsersInTeam` in line below 
https://github.com/mattermost/mattermost-webapp/blob/7d527c9c852f9e931cf253c16f663371340738d5/components/search/types.ts#L54

**Review by sinansonmez — CHANGES_REQUESTED** [review]
Thanks for the effort @babinderrathi I just have one suggestion to eliminate the usage of `any`

**Review by sinansonmez — APPROVED** [review]
LGTM, thanks for the effort @babinderrathi

File: components/search/types.ts [hunk 3]
```diff
@@ -4,12 +4,13 @@
...
 import {Action} from 'redux';
 
-import {ActionFunc, DispatchFunc} from 'mattermost-redux/types/actions';
+import {ActionFunc} from 'mattermost-redux/types/actions';
 
 import {Channel} from '@mattermost/types/channels';
 import {ServerError} from '@mattermost/types/errors';
 
 import {SearchType} from 'types/store/rhs';
+import {UserAutocomplete} from '@mattermost/types/autocomplete';
```
Comment (M-ZubairAhmed) [inline components/search/types.ts:13]: nit
```suggestion
import {Channel} from '@mattermost/types/channels';
import {ServerError} from '@mattermost/types/errors';
import {UserAutocomplete} from '@mattermost/types/autocomplete';

import {SearchType} from 'types/store/rhs';

```

File: components/suggestion/search_user_provider.tsx [hunk 4]
```diff
@@ -75,20 +80,21 @@ class SearchUserSuggestion extends Suggestion {
 }
 
 export default class SearchUserProvider extends Provider {
-    constructor(userSearchFunc) {
+    autocompleteUsersInTeam: (username: string) => Promise<UserAutocomplete>;
```
Comment (M-ZubairAhmed) [inline components/suggestion/search_user_provider.tsx:83]: nit : 
```suggestion
    private autocompleteUsersInTeam: (username: string) => Promise<UserAutocomplete>;
```

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: components/suggestion/search_user_provider.tsx [hunk 5]
```diff
@@ -8,10 +8,15 @@ import BotBadge from 'components/widgets/badges/bot_badge';
...
 
+import {UserProfile} from '@mattermost/types/users';
+import {UserAutocomplete} from '@mattermost/types/autocomplete';
+
 import Provider from './provider';
 import Suggestion from './suggestion.jsx';
+import {ProviderResults} from './generic_user_provider';
 
 class SearchUserSuggestion extends Suggestion {
+    node?: HTMLDivElement | null;
```
Comment (M-ZubairAhmed) [inline components/suggestion/search_user_provider.tsx:19]: nit :
```suggestion
    private node?: HTMLDivElement | null;
```

**Review by M-ZubairAhmed — APPROVED** [review]
💯 It looks good to me. Thank you 🎉

**Pull request merged**